dali/util/cufile.h

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_UTIL_CUFILE_H_
+#define DALI_UTIL_CUFILE_H_
+
+#include <cstdio>
+#include <string>
+#include <memory>
+
+#include "dali/core/api_helper.h"
+#include "dali/core/common.h"
+
+namespace dali {
+
+class DLL_PUBLIC CUFileStream {
+ public:
+  static std::unique_ptr<CUFileStream> Open(const std::string& uri, bool read_ahead,
+                                            bool use_mmap);
+
+  virtual void Close() = 0;
+  virtual size_t Read(uint8_t * buffer, size_t n_bytes, size_t offset = 0) = 0;
+  virtual size_t ReadCPU(uint8_t * buffer, size_t n_bytes) = 0;
```

Comment:
As I understand the goal was to seek file for GPU and CPU, while you read the content on the CPU/GPU separately. 

> get a regular FileStream from the CUFileStream

It would require sharing the state between both of them what could be misused I guess. No idea this is the way we want to do it.
@azrael417 ?

Reply:
Yeah, using any of the reads advances the pos_ inside, so it's not like opening another stream, it's reading on CPU or GPU using the same reader "head".

Still, making it other way round would be easier wrt to the pointer to member and the CuFileStream could be a regular FileStream at the same time, couldn't it?

Reply:
Done, I hope.